go/exosocket/index.go

```diff
@@ -0,0 +1,106 @@
+package exosocket
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/Originate/exocom/go/structs"
+	"github.com/Originate/exocom/go/utils"
+	uuid "github.com/satori/go.uuid"
+
+	"github.com/gorilla/websocket"
+	"github.com/pkg/errors"
+)
+
+// ExoSocket is the low level API to Exocom
+type ExoSocket struct {
+	config                      Config
+	shouldReconnectOnDisconnect bool
+	socket                      *websocket.Conn
+	messageChannel              chan structs.Message
+}
+
+// NewExoSocket returns an ExoSocket with the given config
+func NewExoSocket(config Config) *ExoSocket {
+	return &ExoSocket{config: config}
+}
+
+// Connect brings an ExoSocket instance online
+func (e *ExoSocket) Connect() error {
+	if e.config.Port == 0 {
+		e.config.Port = 80
+	}
+	exocomURL := fmt.Sprintf("ws://%s:%d/services", e.config.Host, e.config.Port)
+	socket, err := utils.ConnectWithRetry(exocomURL, 100)
+	if err != nil {
+		return err
+	}
+	e.shouldReconnectOnDisconnect = true
+	e.socket = socket
+	e.messageChannel = make(chan structs.Message)
```

Comment:
How do you choose a buffer size? I have all the buffers being size one to synchronize them and feels like it has been simpler to debug since reads and write block. Makes the receiver do any longer processing in a separate goroutine.

Reply:
Buffer size would depend on the expected throughput of the system, timeouts, memory constraints etc.

The default size of a channel is 0 (`make(chan structs.Message)` is equivalent to `make(chan structs.Message, 0)`). As per the [spec](https://golang.org/ref/spec#Send_statements) this means a receiver must be ready for any send to proceed.

It's probably a good idea to have a few second's worth of buffering here